src/program/lwaftr/tests/propbased/genyang.lua

```diff
@@ -0,0 +1,369 @@
+module(..., package.seeall)
+
+-- This module provides functions for generating snabb config
+-- commands with random path queries and values
+
+local schema = require("lib.yang.schema")
+
+local capabilities = {['ietf-softwire']={feature={'binding', 'br'}}}
+require('lib.yang.schema').set_default_capabilities(capabilities)
+
+local schemas = { "ietf-softwire", "snabb-softwire-v1" }
+
+-- Generate a get/set command string given a pid string and optional schema
+function generate_get_or_set(pid, schema)
+   local r = math.random()
+   if r > 0.5 then
+      local query, schema = generate_config_xpath(schema)
+      return string.format("./snabb config get -s %s %s \"%s\"", schema, pid, query)
+   else
+      local query, val, schema = generate_config_xpath_and_val(schema)
+      return string.format("./snabb config set -s %s %s \"%s\" \"%s\"",
+                           schema, pid, query, val)
+   end
+end
+
+-- Generate a get command string given a pid string and optional schema/query
+function generate_get(pid, schema, query)
+   if not query then
+      query, schema = generate_config_xpath(schema)
+   end
+   return string.format("./snabb config get -s %s %s \"%s\"", schema, pid, query)
+end
+
+-- Like generate_get but for state queries
+function generate_get_state(pid, schema, query)
+   if not query then
+      query, schema = generate_config_xpath_state(schema)
+   end
+   return string.format("./snabb config get-state -s %s %s \"%s\"", schema, pid, query)
+end
+
+-- Used primarily for repeating a set with a value seen before from a get
+function generate_set(pid, schema, query, val)
+   return string.format("./snabb config set -s %s %s \"%s\" \"%s\"",
+                        schema, pid, query, val)
+end
+
+function run_yang(yang_cmd)
+   local f = io.popen(yang_cmd)
+   local result = f:read("*a")
+   f:close()
+   return result
+end
+
+-- choose an element of an array randomly
+local function choose(choices)
+   local idx = math.random(#choices)
+   return choices[idx]
+end
+
+-- choose a natural number (e.g., index or length of array) by
+-- repeating a cointoss
+local function choose_nat()
+   local r = math.random()
+
+   local function flip(next)
+      local r = math.random()
+      if r < 0.5 then
+         return next
+      else
+         return flip(next + 1)
+      end
+   end
+
+   -- evenly weight first two
+   if r < 0.5 then
+      return choose({1, 2})
+   else
+      return flip(3)
+   end
+end
+
+local function random_hex()
+  return string.format("%x", math.random(0, 15))
+end
+
+local function random_hexes()
+   local str = ""
+   for i=1, 4 do
+      str = str .. random_hex()
+   end
+   return str
+end
+
+-- return a random number, preferring boundary values
+local function choose_range(lo, hi)
```

Comment:
Preferring boundary values is good, but a weighting of 0.5 is probably a bit heavy - how would you feel about 0.1 or 0.2? I'm also inclined to include 0 in the list of numbers to test often.